src/Compression/CompressionCodecSZ3.cpp

```diff
@@ -0,0 +1,307 @@
+            }
+            catch (...)
+            {
+                throw Exception(ErrorCodes::CORRUPTED_DATA, "Unexpected data to compress");
+            }
+            break;
+        }
+        default:
+            throw Exception(ErrorCodes::CORRUPTED_DATA, "Unexpected float width in SZ3 compressed data");
+    }
+
+    size_t offset = 0;
+    memcpy(dest + offset, &float_width, sizeof(UInt8));
+    offset += sizeof(UInt8);
+
+    memcpy(dest + offset, compressed.get(), compressed_size);
+    return static_cast<UInt32>(offset + compressed_size);
+}
+
+void CompressionCodecSZ3::setDimensions(const std::vector<size_t> & dimensions_)
+{
+    dimensions = dimensions_;
+}
+
+void CompressionCodecSZ3::doDecompressData(const char * source, UInt32 source_size, char * dest, UInt32 /*uncompressed_size*/) const
+{
+    UInt8 width = static_cast<UInt8>(*source);
+    --source_size;
+    ++source;
+
+    SZ3::Config config;
+    switch (width)
+    {
+        case 4: {
+            try
+            {
+                float * dest_typed = reinterpret_cast<float *>(dest);
+                SZ_decompress<float>(config, const_cast<char *>(source), source_size, dest_typed);
+            }
+            catch (...)
+            {
+                throw Exception(ErrorCodes::CORRUPTED_DATA, "Unexpected data to compress");
+            }
+            break;
+        }
+        case 8: {
+            try
+            {
+                double * dest_typed = reinterpret_cast<double *>(dest);
+                SZ_decompress<double>(config, const_cast<char *>(source), source_size, dest_typed);
+            }
+            catch (...)
+            {
+                throw Exception(ErrorCodes::CORRUPTED_DATA, "Unexpected data to compress");
+            }
+            break;
+        }
+        default:
+            throw Exception(ErrorCodes::CORRUPTED_DATA, "Unexpected float width in SZ3 compressed data");
+    }
+}
+
+UInt8 getFloatByteWidth(const IDataType & column_type)
+{
+    if (!WhichDataType(column_type).isNativeFloat())
+        throw Exception(
+                ErrorCodes::BAD_ARGUMENTS,
+                "Codec 'SZ3' is not applicable for {} because the data type is not Float*",
+                column_type.getName());
+
+    return column_type.getSizeOfValueInMemory();
+}
+
+SZ3::ALGO getSZ3Algorithm(const String & algorithm)
+{
+    return SZ3::ALGO_MAP.at(algorithm);
+}
+
+SZ3::EB getSZ3ErrorBoundMode(const String & error_bound_mode)
+{
+    return SZ3::EB_MAP.at(error_bound_mode);
+}
+
+void registerCodecSZ3(CompressionCodecFactory & factory)
+{
+    auto method_code = static_cast<UInt8>(CompressionMethodByte::SZ3);
+    auto codec_builder = [&](const ASTPtr & arguments, const IDataType * column_type) -> CompressionCodecPtr
+    {
+        UInt8 float_width = 4;
+        if (column_type)
+            float_width = getFloatByteWidth(*column_type);
+
+        bool use_zero_error_bound = !column_type || WhichDataType(column_type).isUInt64();
```

Comment:
Correct me if I am wrong, but I think that we needed the special handling only for Array offset columns (UInt64). Now that you forbid compressing array offsets anyways, we probably can get rid of l. 264?